### PR TITLE
[DEVHAS-183] Use separate Job to run git clone and CDQ analysis out of HAS controller

### PR DESCRIPTION
**Pull request by yangcao77**
### What does this PR do?:

- Adds a separate Go module/docker image for cdq analysis, including git clone and alizer analysis against the repo. 
- A separate Kubernetes job out of HAS controller will be spawned to run cdq analysis
- result will be stored in a config map created by the job
- HAS controller will wait and read the config map for cdq result

### Which issue(s)/story(ies) does this PR fixes:

https://issues.redhat.com/projects/DEVHAS/issues/DEVHAS-183

### PR acceptance criteria:

- [x] Unit/Functional tests

  

- [ ] Documentation 

   

- [ ] Client Impact

### CODE REVIEW

**Comment by flacatus** [comment]
```log
2023-02-02T18:26:14.278Z	ERROR	controllers.Component	Unable to GET devfile.yaml, exiting reconcile loop has-e2e-obom/quarkus-component-e2e-xzzsoqvxca	{"appstudio-component": "HAS", "Component": "has-e2e-obom/quarkus-component-e2e-xzzsoqvxca", "clusterName": "", "error": "Get \"devfile.yaml\": unsupported protocol scheme \"\""}
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile
	/go/pkg/mod/github.com/kcp-dev/controller-runtime@v0.12.2-0.20220808200255-4b60fd66e5de/pkg/internal/controller/controller.go:121
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler
	/go/pkg/mod/github.com/kcp-dev/controller-runtime@v0.12.2-0.20220808200255-4b60fd66e5de/pkg/internal/controller/controller.go:320
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem
	/go/pkg/mod/github.com/kcp-dev/controller-runtime@v0.12.2-0.20220808200255-4b60fd66e5de/pkg/internal/controller/controller.go:273
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2
	/go/pkg/mod/github.com/kcp-dev/controller-runtime@v0.12.2-0.20220808200255-4b60fd66e5de/pkg/internal/controller/controller.go:234
2023-02-02T18:26:14.278Z	INFO	controllers.Component	API Resource change event failed: Create, error: unable to GET from devfile.yaml	{"appstudio-component": "HAS", "Component": "has-e2e-obom/quarkus-component-e2e-xzzsoqvxca", "clusterName": "", "audit": "true", "name": "quarkus-component-e2e-xzzsoqvxca", "resource": "Component", "action": "Create"}

```

I see this log multiple times

File: cdq-analysis/pkg/util.go [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+
+// CurlEndpoint curls the endpoint and returns the response or an error if the response is a non-200 status
+func CurlEndpoint(endpoint string) ([]byte, error) {
+	var respBytes []byte
+	/* #nosec G107 --  The URL is validated by the CDQ if the request is coming from the UI.  If we do happen to download invalid bytes, the devfile parser will catch this and fail. */
+	resp, err := http.Get(endpoint)
+	if err != nil {
+		return nil, err
+	}
+	defer resp.Body.Close()
```
[the remark this answers is not in the document]
Comment (yangcao77) [reply cdq-analysis/pkg/util.go:82]: unrelated to this PR change

File: cdq-analysis/pkg/util.go [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+	"github.com/devfile/registry-support/index/generator/schema"
+	registryLibrary "github.com/devfile/registry-support/registry-library/library"
+	"github.com/redhat-developer/alizer/go/pkg/apis/model"
+)
+
+// CloneRepo clones the repoURL to clonePath
+func CloneRepo(clonePath, repoURL string, token string) error {
+	exist, err := IsExist(clonePath)
+	if !exist || err != nil {
+		os.MkdirAll(clonePath, 0750)
```
[the remark this answers is not in the document]
Comment (yangcao77) [reply cdq-analysis/pkg/util.go:39]: unrelated to this PR change

File: controllers/componentdetectionquery_controller.go [hunk 3]
```diff
@@ -161,69 +167,83 @@
...
+				log.Error(err, fmt.Sprintf("Error waiting for configmap creation ...%v", req.NamespacedName))
+				r.SetCompleteConditionAndUpdateCR(ctx, req, &componentDetectionQuery, copiedCDQ, err)
+				cleanupK8sResources(log, clientset, ctx, fmt.Sprintf("%s-job", req.Name), req.Name, req.Namespace)
+				return ctrl.Result{}, nil
+			}
+			var devfileMapReturned map[string][]byte
+			var devfilesURLMapReturned map[string]string
+			var dockerfileContextMapReturned map[string]string
+			var errMapReturned map[string]string
+			json.Unmarshal(cm.BinaryData["devfilesMap"], &devfileMapReturned)
```
[the remark this answers is not in the document]
Comment (yangcao77) [reply controllers/componentdetectionquery_controller.go:227]: unrelated to this PR change

File: cdq-analysis/pkg/util.go [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+	}
+	cloneURL := repoURL
+	// Execute does an exec.Command on the specified command
+	if token != "" {
+		tempStr := strings.Split(repoURL, "https://")
+
+		// e.g. https://token:<token>@github.com/owner/repoName.git
+		cloneURL = fmt.Sprintf("https://token:%s@%s", token, tempStr[1])
+	}
+	c := exec.Command("git", "clone", cloneURL, clonePath)
```
[the remark this answers is not in the document]
Comment (yangcao77) [reply cdq-analysis/pkg/util.go:49]: unrelated to this PR change

**Pull request closed without merging**